File: src/api/dcps/sacs/code/DDS/OpenSplice/ContentFilteredTopic.cs
/*
 *                         Vortex OpenSplice
 *
 *   This software and documentation are Copyright 2006 to TO_YEAR ADLINK
 *   Technology Limited, its affiliated companies and licensors. All rights
 *   reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 */

using System;
using System.Runtime.InteropServices;
using DDS;
using DDS.OpenSplice.OS;
using DDS.OpenSplice.CustomMarshalers;

namespace DDS.OpenSplice
{
    internal class ContentFilteredTopic : SacsSuperClass, IContentFilteredTopic, ITopicDescriptionImpl
    {
        private DomainParticipant domainParticipant;
        private TypeSupport typeSupport;
        private string typeName;
        private string topicName;
        private string filterExpression;
        private string[] filterParameters;
        private string topicExpression;
        private uint nrUsers = 0;
        private Topic relatedTopic;

        internal ContentFilteredTopic(
                string typeName,
                string topicName,
                DomainParticipant domainParticipant,
                string filterExpression,
                string[] filterParameters)
        {
            this.typeName = typeName;
            this.topicName = topicName;
            this.domainParticipant = domainParticipant;
            this.filterExpression = filterExpression;
            this.filterParameters = filterParameters;
        }

        internal ReturnCode init(Topic relatedTopic)
        {
            ReturnCode result;

            MyDomainId = relatedTopic.MyDomainId;

            this.relatedTopic = relatedTopic;
            lock(relatedTopic)
            {
                if (relatedTopic.rlReq_isAlive)
                {
                    result = relatedTopic.ValidateFilter(filterExpression, filterParameters);
                    if (result == DDS.ReturnCode.Ok)
                    {
                        ITopicDescriptionImpl relatedTopicDescr = relatedTopic as ITopicDescriptionImpl;
                        relatedTopicDescr.wlReq_IncrNrUsers();
                        this.topicExpression = relatedTopicDescr.rlReq_TopicExpression + " where " + this.filterExpression;
                        this.typeSupport = relatedTopicDescr.rlReq_TypeSupport;
                        result = base.init(IntPtr.Zero, false);
                    }
                }
                else
                {
                    result = DDS.ReturnCode.BadParameter;
                    ReportStack.Report(result, "ContentFilteredTopic \"" + topicName + "\" is referring to an already deleted topic.");
                }
            }
            return result;
        }

        internal override ReturnCode wlReq_deinit()
        {
            ReturnCode result;

            if (nrUsers > 0)
            {
                result = DDS.ReturnCode.PreconditionNotMet;
                ReportStack.Report(result, "ContentFilteredTopic \"" + topicName + "\" still in use by " + nrUsers + " Writers/Readers.");
            }
            else
            {
                result = base.wlReq_deinit();
                if (result == DDS.ReturnCode.Ok)
                {
                    result = (relatedTopic as ITopicDescriptionImpl).DecrNrUsers();
                }
            }
            return result;
        }

        /// <summary>
        /// This property returns the registered name of the data type associated with the TopicDescription
        /// (inherited from TopicDescription)
        /// </summary>
        public string TypeName
        {
            get
            {
                bool isAlive;
                string name = null;

                ReportStack.Start();
                lock(this)
                {
                    isAlive = this.rlReq_isAlive;
                    if (isAlive)
                    {
                        name = typeName;
                    }
                }
                ReportStack.Flush(this, !isAlive);

                return name;
            }
        }

        /// <summary>
        /// This property returns the name used to create the TopicDescription.
        /// (inherited from TopicDescription)
        /// </summary>
        public string Name
        {
            get
            {
                bool isAlive;
                string name = null;

                ReportStack.Start();
                lock(this)
                {
                    isAlive = this.rlReq_isAlive;
                    if (isAlive)
                    {
                        name = topicName;
                    }
                }
                ReportStack.Flush(this, !isAlive);

                return name;
            }
        }

        /// <summary>
        /// This property returns the DomainParticipant associated with the TopicDescription or
        /// a null DomainParticipant.
        /// (inherited from TopicDescription)
        /// </summary>
        public IDomainParticipant Participant
        {
            get
            {
                bool isAlive;
                DomainParticipant participant = null;

                ReportStack.Start();
                lock(this)
                {
                    isAlive = this.rlReq_isAlive;
                    if (isAlive)
                    {
                        participant = domainParticipant;
                    }
                }
                ReportStack.Flush(this, !isAlive);

                return participant;
            }
        }

        public string GetFilterExpression()
        {
            bool isAlive;
            string fexpr = null;

            ReportStack.Start();
            lock(this)
            {
                isAlive = this.rlReq_isAlive;
                if (isAlive)
                {
                    fexpr = filterExpression;
                }
            }
            ReportStack.Flush(this, !isAlive);

            return fexpr;
        }

        public ReturnCode GetExpressionParameters(ref string[] expressionParameters)
        {
            expressionParameters = null;
            ReturnCode result = DDS.ReturnCode.AlreadyDeleted;

            ReportStack.Start();
            lock(this)
            {
                if (this.rlReq_isAlive)
                {
                    expressionParameters = this.filterParameters;
                    result = DDS.ReturnCode.Ok;
                }
            }
            ReportStack.Flush(this, result != DDS.ReturnCode.Ok);

            return result;
        }

        public ReturnCode SetExpressionParameters(params string[] expressionParameters)
        {
            ReturnCode result = DDS.ReturnCode.Unsupported;
            ReportStack.Start();

//            using (SequenceStringToArrMarshaler marshaler = new SequenceStringToArrMarshaler())
//            {
//                result = WriteLock();
//                if (result == DDS.ReturnCode.Ok)
//                {
//                    result = marshaler.CopyIn(expressionParameters);
//                    if (result == ReturnCode.Ok)
//                    {
//                        result = uResultToReturnCode(
//                                User.ContentFilteredTopic.Set(rlReq_UserPeer, marshaler.UserPtr, (uint) expressionParameters.Length));
//                        if (result == ReturnCode.Ok)
//                        {
//                            marshaler.CopyOut(ref this.filterParameters); // Make deep copy.
//                        }
//                    }
//                    WriteUnlock();
//                }
//            }
            ReportStack.Report(result, "Operation not yet supported.");
            ReportStack.Flush(this, result != ReturnCode.Ok);

            return result;
        }

        public ITopic RelatedTopic
        {
            get
            {
                bool isAlive;
                Topic topic = null;

                ReportStack.Start();
                lock(this)
                {
                    isAlive = this.rlReq_isAlive;
                    if (isAlive)
                    {
                        topic = relatedTopic;
                    }
                }
                ReportStack.Flush(this, !isAlive);

                return topic;
            }
        }

        string ITopicDescriptionImpl.rlReq_TopicExpression
        {
            get
            {
                return topicExpression;
            }
        }

        string[] ITopicDescriptionImpl.rlReq_TopicExpressionParameters
        {
            get
            {
                return filterParameters;
            }
        }

        void ITopicDescriptionImpl.wlReq_IncrNrUsers()
        {
            nrUsers++;
        }

        ReturnCode ITopicDescriptionImpl.DecrNrUsers()
        {
            ReturnCode result = DDS.ReturnCode.AlreadyDeleted;

            lock(this)
            {
                if (this.rlReq_isAlive)
                {
                    nrUsers--;
                    result = DDS.ReturnCode.Ok;
                }
            }
            return result;
        }

        TypeSupport ITopicDescriptionImpl.rlReq_TypeSupport
        {
            get
            {
                return typeSupport;
            }
            set
            {
                typeSupport = value;
            }
        }
    }
}

File: src/api/dcps/c++/common/code/ContentFilteredTopic.cpp
/*
 *                         OpenSplice DDS
 *
 *   This software and documentation are Copyright 2006 to TO_YEAR PrismTech
 *   Limited, its affiliated companies and licensors. All rights reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 *
 */
#include "ContentFilteredTopic.h"
#include "DomainParticipant.h"
#include "TopicDescription.h"
#include "TypeSupportMetaHolder.h"
#include "ReportUtils.h"

DDS::OpenSplice::ContentFilteredTopic::ContentFilteredTopic() :
    DDS::OpenSplice::CppSuperClass(DDS::OpenSplice::CONTENTFILTEREDTOPIC),
    relatedTopic(NULL)
{
}

DDS::ReturnCode_t
DDS::OpenSplice::ContentFilteredTopic::init(
    DDS::OpenSplice::DomainParticipant *participant,
    const DDS::Char *topic_name,
    DDS::OpenSplice::Topic *related_topic,
    const DDS::Char *filter_expression,
    const DDS::StringSeq &filter_parameters)
{
    return this->nlReq_init(participant, topic_name, related_topic, filter_expression, filter_parameters);
}

DDS::ReturnCode_t
DDS::OpenSplice::ContentFilteredTopic::nlReq_init(
    DDS::OpenSplice::DomainParticipant *participant,
    const DDS::Char *topic_name,
    DDS::OpenSplice::Topic *related_topic,
    const DDS::Char *filter_expression,
    const DDS::StringSeq &filter_parameters)
{
    DDS::ReturnCode_t result;
    DDS::OpenSplice::TypeSupportMetaHolder *type_holder = NULL;
    DDS::Char *expression = NULL;
    const char *related_topic_name = NULL;
    const char *related_type_name = NULL;
    DDS::ULong length;
    const DDS::Char *format = "select * from %s where %s";

    assert(participant);
    assert(topic_name);
    assert(related_topic);
    assert(filter_expression);

    result = DDS::OpenSplice::CppSuperClass::nlReq_init();
    if (result == DDS::RETCODE_OK) {
        result = related_topic->write_lock();
        if (result == DDS::RETCODE_OK) {
            type_holder        = related_topic->tsMetaHolder;
            related_type_name  = related_topic->type_name.in();
            related_topic_name = related_topic->topic_name.in();
            result = related_topic->validate_filter(filter_expression, filter_parameters);
            if (result == DDS::RETCODE_OK) {
                length = strlen(format) + strlen(related_topic_name) + strlen(filter_expression) + 1;
                expression = DDS::string_alloc(length);
                if (expression) {
                    snprintf(expression, length, format, related_topic_name, filter_expression);
                    result = DDS::OpenSplice::TopicDescription::nlReq_init(participant,
                                                                      topic_name,
                                                                      related_type_name,
                                                                      expression,
                                                                      type_holder);
                    DDS::string_free(expression);
                    if (result == DDS::RETCODE_OK) {
                        (void) DDS::Topic::_duplicate(related_topic);
                        relatedTopic = related_topic;
                        filterExpression = DDS::string_dup(filter_expression);
                        filterParameters = filter_parameters;
                        related_topic->wlReq_incrNrUsers();
                        setDomainId(relatedTopic->getDomainId());
                    }
                } else {
                    result = DDS::RETCODE_OUT_OF_RESOURCES;
                    CPP_REPORT(result, "Could not allocate memory.");
                }
            }
            related_topic->unlock();
        }
    }

    return result;
}

DDS::ReturnCode_t
DDS::OpenSplice::ContentFilteredTopic::wlReq_deinit()
{
    DDS::ReturnCode_t result = DDS::RETCODE_PRECONDITION_NOT_MET;

    if (this->rlReq_getNrUsers() == 0) {
        result = this->relatedTopic->write_lock();
        if (result == DDS::RETCODE_OK) {
            this->relatedTopic->wlReq_decrNrUsers();
            this->relatedTopic->unlock();
            DDS::release(this->relatedTopic);
            this->relatedTopic = NULL;

            result = DDS::OpenSplice::TopicDescription::wlReq_deinit();
            if (result == DDS::RETCODE_OK) {
                result = DDS::OpenSplice::CppSuperClass::wlReq_deinit();
            }
        }
    } else {
        CPP_REPORT(result, "ContentFilteredTopic still in use.");
    }
    return result;
}

DDS::OpenSplice::ContentFilteredTopic::~ContentFilteredTopic()
{
}

char *
DDS::OpenSplice::ContentFilteredTopic::get_filter_expression (
) THROW_ORB_EXCEPTIONS
{
    DDS::ReturnCode_t result;
    char *expression = NULL;

    CPP_REPORT_STACK();

    result = this->read_lock ();
    if (result == DDS::RETCODE_OK) {
        expression = DDS::string_dup(filterExpression);
        if (expression == NULL) {
            result = DDS::RETCODE_OUT_OF_RESOURCES;
            CPP_REPORT(result, "Could not copy filter expression.");
        }
        this->unlock ();
    }

    CPP_REPORT_FLUSH(this, result != DDS::RETCODE_OK);

    return expression;
}

DDS::ReturnCode_t
DDS::OpenSplice::ContentFilteredTopic::get_expression_parameters (
    ::DDS::StringSeq & expression_parameters
) THROW_ORB_EXCEPTIONS
{
    DDS::ReturnCode_t result;

    CPP_REPORT_STACK();

    result = this->read_lock ();
    if (result == DDS::RETCODE_OK) {
        expression_parameters = filterParameters;
        this->unlock ();
    }

    CPP_REPORT_FLUSH(this, result != DDS::RETCODE_OK);

    return result;
}

DDS::ReturnCode_t
DDS::OpenSplice::ContentFilteredTopic::set_expression_parameters (
    const ::DDS::StringSeq & expression_parameters
) THROW_ORB_EXCEPTIONS
{
    // We currently have no user layer object to forward the request to.
    OS_UNUSED_ARG(expression_parameters);
    DDS::ReturnCode_t result = DDS::RETCODE_UNSUPPORTED;

    CPP_REPORT_STACK();

    CPP_REPORT(result, "Operation not yet supported.");

    CPP_REPORT_FLUSH(this, result != DDS::RETCODE_OK);

    return result;
}

DDS::Topic_ptr
DDS::OpenSplice::ContentFilteredTopic::get_related_topic (
) THROW_ORB_EXCEPTIONS
{
    DDS::ReturnCode_t result;
    DDS::Topic_ptr related_topic = NULL;

    CPP_REPORT_STACK();

    result = this->check ();
    if (result == DDS::RETCODE_OK) {
        related_topic = DDS::Topic::_duplicate(relatedTopic);
    }

    CPP_REPORT_FLUSH(this, result != DDS::RETCODE_OK);

    return related_topic;
}
